wezterm-gui/src/termwindow/mod.rs

```diff
@@ -1018,44 +1018,124 @@ impl TermWindow {
                     Some(pane) => pane,
                     None => return Ok(true),
                 };
-                pane.send_paste(text.as_str())?;
+                let txt = self.user_dropped_event( "user-dropped-string", vec![text.clone()]);
+                pane.send_paste(txt.as_deref().unwrap_or(&text))?;
                 Ok(true)
             }
             WindowEvent::DroppedUrl(urls) => {
                 let pane = match self.get_active_pane_or_overlay() {
                     Some(pane) => pane,
                     None => return Ok(true),
                 };
-                let urls = urls
-                    .iter()
-                    .map(|url| self.config.quote_dropped_files.escape(&url.to_string()))
-                    .collect::<Vec<_>>()
-                    .join(" ")
-                    + " ";
-                pane.send_paste(urls.as_str())?;
+                let urls_string = self.user_dropped_event( 
+                    "user-dropped-urls",
+                    urls.iter().map(|url| url.to_string()).collect::<Vec<_>>(),
+                );
+                pane.send_paste(
+                    urls_string.as_deref().unwrap_or(
+                        &urls
+                            .iter()
+                            .map(|url| url.to_string())
+                            .collect::<Vec<_>>()
+                            .join(" "),
+                    ),
+                )?;
```

Comment:
re: async, the way to tackle this is to spawn the processing into a task via `spawn_into_main_thread`. You'd copy/clone the data you need and spawn the task. It would make processing these events always be async wrt. to the GUI level drop event, but I think that's fine.

Look for other examples of this kind of spawning in wezterm-gui; I don't know if there is an ideal example to use as a role model, but hopefully there's enough there to get you going.

Reply:
The reason I had gone with sync originally was because it seemed the callbacks which are expected to return something in the course of UI and user interactions worked this way. Specifically `format-tab-title` and I believe there was really only one other. Also because the lua evaluations and events use many different run, spawn and do-on-main functions which I had not fully grasped why and when and what they were best utilized for. 

It was a tad confusing and not as well documented as far as understanding why X was used for Y. But I also blame myself in not truly understanding the rusts async to begin with. With how the futures are handled and terminology in the smol/tokio runtimes.

Reply:
The async handling within rust and its futures are a bit confusing to me currently. BUt it also may be trying to incorporate the current Wezterm API of `promise`, and `config::with_lua*`.

THat being said, I did try my best to extract the common stuff into a single function, and the whole process handles in an async closure, including the `pane:send_paste` since my limited experience with the futures and async thread/send stuff being a bit out of my league. However, it runs perfectly, but I do not know if there are any issues that may arise out of this implementation that I am not aware of.